enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/transaction/log/checkpoint/ConfigurableIOLimiter.java

```diff
@@ -32,41 +32,82 @@
 
 public class ConfigurableIOLimiter implements IOLimiter
 {
-    private static final AtomicIntegerFieldUpdater<ConfigurableIOLimiter> disableCountUpdater =
-            AtomicIntegerFieldUpdater.newUpdater( ConfigurableIOLimiter.class, "disabledCount" );
+    private static final AtomicLongFieldUpdater<ConfigurableIOLimiter> stateUpdater =
+            AtomicLongFieldUpdater.newUpdater( ConfigurableIOLimiter.class, "state" );
 
     private static final int NO_LIMIT = 0;
     private static final int QUANTUM_MILLIS = 100;
     private static final int TIME_BITS = 32;
     private static final long TIME_MASK = (1L << TIME_BITS) - 1;
+    private static final int QUANTUMS_PER_SECOND = (int) (TimeUnit.SECONDS.toMillis( 1 ) / QUANTUM_MILLIS);
 
-    private final int iopq; // IOs per quantum
     private final ObjLongConsumer<Object> pauseNanos;
 
-    @SuppressWarnings( "unused" ) // Updated via disableCountUpdater
-    private volatile int disabledCount;
+    /**
+     * Upper 32 bits is the "disabled counter", lower 32 bits is the "IOs per quantum" field.
+     * The "disabled counter" is modified online in 2-increments, leaving the lowest bit for signalling when
+     * the limiter disabled by configuration.
+     */
+    @SuppressWarnings( "unused" ) // Updated via stateUpdater
+    private volatile long state;
 
     public ConfigurableIOLimiter( Config config )
     {
         this( config, LockSupport::parkNanos );
     }
 
-    // Only visible for testing
+    // Only visible for testing.
     ConfigurableIOLimiter( Config config, ObjLongConsumer<Object> pauseNanos )
     {
         this.pauseNanos = pauseNanos;
-        int quantumsPerSecond = (int) (TimeUnit.SECONDS.toMillis( 1 ) / QUANTUM_MILLIS);
         Integer iops = config.get( GraphDatabaseSettings.check_point_iops_limit );
+        updateConfiguration( iops );
+        config.registerDynamicUpdateListener( GraphDatabaseSettings.check_point_iops_limit,
+                ( prev, update ) -> updateConfiguration( update ) );
+    }
+
+    private void updateConfiguration( Integer iops )
+    {
+        long oldState;
+        long newState;
         if ( iops == null || iops < 1 )
         {
-            iopq = NO_LIMIT;
-            disabledCount = 1; // This permanently disables IO limiting
+            do
+            {
+                oldState = stateUpdater.get( this );
+                int disabledCounter = getDisabledCounter( state );
```

Comment:
I think it would be cleaner to derive the new value from the `oldState` just retrieved. But that's just a matter of taste, feel free to ignore.

Reply:
Good catch. It was not supposed to be like that.